### PR TITLE
EVG-2686 implement server logic for reverting an event

### PR DESCRIPTION
**Pull request by john-m-liu**

### CODE REVIEW

File: model/event/admin_event.go [hunk 1]
```diff
@@ -138,3 +138,26 @@ func convertRaw(in rawAdminEventData) (*AdminEventData, error) {
...
 }
+
+// RevertConfig reverts one config section to the before state of the specified GUID in the event log
+func RevertConfig(guid string, user string) error {
+	events, err := FindAdmin(ByGuid(guid))
+	if err != nil {
+		return errors.Wrap(err, "problem finding events")
+	}
+	if len(events) == 0 {
+		return fmt.Errorf("unable to find event with GUID %s", guid)
```
Comment (bsamek) [inline model/event/admin_event.go]: `errors.Errorf`

File: model/event/admin_event.go [hunk 2]
```diff
@@ -138,3 +138,26 @@ func convertRaw(in rawAdminEventData) (*AdminEventData, error) {
...
+	current := evergreen.ConfigRegistry.GetSection(data.Section)
+	if current == nil {
+		return fmt.Errorf("unable to find section %s", data.Section)
+	}
+	err = data.Changes.Before.Set()
+	if err != nil {
+		return errors.Wrap(err, "problem updating settings")
+	}
+
+	return LogAdminEvent(data.Section, current, data.Changes.Before, user)
```
Comment (bsamek) [inline model/event/admin_event.go]: Does `data.Changes.Before` need to be scrubbed, or does this happen later?

File: rest/model/admin.go [hunk 3]
```diff
@@ -118,20 +118,38 @@ func (as *APIAdminSettings) BuildFromService(h interface{}) error {
...
-		Plugins:            evergreen.PluginConfig{},
-		PprofPort:          *as.PprofPort,
-		SuperUsers:         as.SuperUsers,
+		Credentials: map[string]string{},
+		Expansions:  map[string]string{},
+		Keys:        map[string]string{},
+		Plugins:     evergreen.PluginConfig{},
+		SuperUsers:  as.SuperUsers,
+	}
+	if as.ApiUrl != nil {
```
Comment (bsamek) [inline rest/model/admin.go]: Is there a risk any of these might be nil pointers?

File: model/event/admin_event.go (hunk 2 above)
Comment (richardsamuels) [reply model/event/admin_event.go]: I think scrubbing has been removed completely.

File: model/event/admin_event.go [hunk 4]
```diff
@@ -138,3 +138,26 @@ func convertRaw(in rawAdminEventData) (*AdminEventData, error) {
...
+		return errors.Wrap(err, "problem finding events")
+	}
+	if len(events) == 0 {
+		return fmt.Errorf("unable to find event with GUID %s", guid)
+	}
+	evt := events[0]
+	data := evt.Data.Data.(*AdminEventData)
+	current := evergreen.ConfigRegistry.GetSection(data.Section)
+	if current == nil {
+		return fmt.Errorf("unable to find section %s", data.Section)
```
Comment (richardsamuels) [inline model/event/admin_event.go]: errors.Errorf

File: model/event/admin_event.go (hunk 1 above)
Comment (john-m-liu) [reply model/event/admin_event.go]: Both are nearly the same thing, but it looks like errors.Errorf has a stack (which isn't needed here)

File: rest/model/admin.go (hunk 3 above)
Comment (richardsamuels) [inline rest/model/admin.go]: Side note, I keep having to do this a lot. Should we just be using recover() for this sort of thing instead?

File: rest/route/admin_routes.go [hunk 5]
```diff
@@ -339,3 +339,53 @@ func (h *restartHandler) Execute(ctx context.Context, sc data.Connector) (Respon
...
+}
+
+func (h *revertHandler) ParseAndValidate(ctx context.Context, r *http.Request) error {
+	if err := util.ReadJSONInto(r.Body, h); err != nil {
+		return err
+	}
+	defer r.Body.Close()
+	if h.GUID == "" {
+		return rest.APIError{
+			StatusCode: 400,
```
Comment (richardsamuels) [inline rest/route/admin_routes.go]: http.StatusBadRequest

File: model/event/admin_event.go (hunk 2 above)
Comment (john-m-liu) [reply model/event/admin_event.go]: Yep, we're return the data verbatim now

File: rest/route/admin_routes.go [hunk 6]
```diff
@@ -339,3 +339,53 @@ func (h *restartHandler) Execute(ctx context.Context, sc data.Connector) (Respon
...
+
+func (h *revertHandler) Handler() RequestHandler {
+	return &revertHandler{}
+}
+
+func (h *revertHandler) ParseAndValidate(ctx context.Context, r *http.Request) error {
+	if err := util.ReadJSONInto(r.Body, h); err != nil {
+		return err
+	}
+	defer r.Body.Close()
```
Comment (richardsamuels) [inline rest/route/admin_routes.go]: I don't usually close bodies in ParseAndValidate. I think this is handled by a higher level.

File: model/event/admin_event.go (hunk 4 above)
Comment (john-m-liu) [reply model/event/admin_event.go]: Both are equivalent

File: rest/model/admin.go (hunk 3 above)
Comment (john-m-liu) [reply rest/model/admin.go]: Personally I don't use an error handler unless I can't anticipate the types of errors that may be encountered. I think this is more robust

File: rest/route/admin_routes.go (hunk 6 above)
Comment (john-m-liu) [reply rest/route/admin_routes.go]: I took a look at the implementation in handler.go and didn't see that this is closed

File: rest/route/admin_routes.go (hunk 5 above)
Comment (john-m-liu) [reply rest/route/admin_routes.go]: fixed

File: rest/model/admin.go (hunk 3 above)
Comment (richardsamuels) [reply rest/model/admin.go]: Fair point. I find it super tedious when interacting with REST apis.

File: model/event/event_finder.go [hunk 7]
```diff
@@ -120,6 +120,12 @@ func RecentAdminEvents(n int) db.Q {
 	}).Sort([]string{"-" + TimestampKey}).Limit(n)
 }
 
+func ByGuid(guid string) db.Q {
+	return db.Query(bson.M{
+		DataKey + "." + "guid": guid,
```
Comment (richardsamuels) [inline model/event/event_finder.go]: bsonutil.GetDottedKeyName

**Pull request merged**